src/server.c

```diff
@@ -856,9 +863,30 @@ int updateClientMemUsage(client *c) {
             c->mem_usage_bucket_node = NULL;
         }
     }
+}
+
+/* This is called only if explicit clients when something changed their buffers,
+ * so we can track clients' memory and enforce clients' maxmemory in real time.
+ *
+ * This also adds the client to the correct memory usage bucket. Each bucket contains
+ * all clients with roughly the same amount of memory. This way we group
+ * together clients consuming about the same amount of memory and can quickly
+ * free them in case we reach maxmemory-clients (client eviction).
+ *
+ * returns 1 if client eviction for this client is allowed, 0 otherwise.
+ */
+int updateClientMemUsageAndBucket(client *c) {
+    serverAssert(io_threads_op == IO_THREADS_OP_IDLE);
+    int allow_eviction = clientEvictionAllowed(c);
+    removeClientFromMemUsageBucket(c, allow_eviction);
+
     if (allow_eviction) {
```

Comment:
If the default config is to keep eviction disabled. We could maybe handle the no eviction condition first.

```
if (!allow_eviction) {
    return 0;
}

```

And then the allow_eviction code flow. WDYT @oranagra ?

Reply:
Yeah, good idea. 
Not because of default config or performance. Just trading extra nesting with an early exit.